### ISSUE TITLE
Rust does not detect `avx` or `avx2`?

### ISSUE THREAD

**Issue by JustForFun88** [rust-lang/rust#115701]
I tried this code:

```rust
use core::arch::x86_64 as x86;
use core::mem;

#[cfg(target_feature = "avx")]
pub(crate) struct GroupIter(x86::__m256i);

fn main() {
    #[cfg(target_feature = "avx")]
    let size = mem::size_of::<GroupIter>();
    #[cfg(target_feature = "avx")]
    assert_eq!(size, 0);
}
```

And then this one:

```rust
use core::arch::x86_64 as x86;
use core::mem;

#[cfg(target_feature = "avx2")]
pub(crate) struct GroupIter(x86::__m512i);

fn main() {
    #[cfg(target_feature = "avx2")]
    let size = mem::size_of::<GroupIter>();
    #[cfg(target_feature = "avx2")]
    assert_eq!(size, 0);
}
```

Finally just code from std (https://doc.rust-lang.org/beta/core/arch/#static-cpu-feature-detection):

```rust
#[cfg(all(
    any(target_arch = "x86", target_arch = "x86_64"),
    target_feature = "avx2"
))]
fn foo(val: usize) {
    panic!("{}", val);
}

#[cfg(not(all(
    any(target_arch = "x86", target_arch = "x86_64"),
    target_feature = "avx2"
)))]
fn foo(val: usize) {
    println!("{}", val);
}

fn main() {
    foo(10);
}
```
I expected **that all functions would panic** (since compilation is done on the `x86_64` architecture), which is what happens if I remove `cfg`: [example 1](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e2db2ca8b1081c32ed864fa6060de20d), [example 2](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ce9a07e6d3eadad50560a2b9a16438de), [example 3](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7c3cc5a24ec335bf87bfe81cedb04639)

Instead, this happened: **It compiles.**
